src/js/common/schemaform/definitions/address.js

```diff
@@ -225,3 +225,289 @@ export function uiSchema(label = 'Address', useStreet3 = false, isRequired = nul
+
+  function validateCity(errors, city, { veteran }) {
+    const state = veteran[addressName]['view:state'];
+    const isMilitaryState = militaryStates.includes(state);
+    if (city && state && isMilitaryState && !isMilitaryCity(city)) {
+      errors.addError(
+        'Please enter APO, FPO, or DPO'
+      );
+    }
+  }
+
+  return {
+    'ui:title': title,
+    type: {
+      'ui:options': {
+        hideIf: () => true,
+        updateSchema: (formData, typeSchema) => {
+          /* eslint-disable no-param-reassign */
+          const address = formData.veteran[addressName];
+          const country = address.country;
+          const state = address['view:state'];
+          const city = address['view:city'];
+          const isDomestic = country === 'USA';
+          const isMilitary = militaryPostOfficeTypeCodes.includes(city) || militaryStates.includes(state);
+          if (isDomestic) {
+            formData.veteran[addressName].type = 'DOMESTIC';
+          } else if (isMilitary) {
+            formData.veteran[addressName].type = 'MILITARY';
+          } else if (country) {
+            formData.veteran[addressName].type = 'INTERNATIONAL';
+          }
+          return typeSchema;
+          /* eslint-enable no-param-reassign */
+        }
+      }
+    },
+    country: {
+      'ui:title': 'Country'
+    },
+    'view:state': {
+      'ui:title': 'State',
+      'ui:required': (formData) => {
+        const address = formData.veteran[addressName];
+        const isUSA = address.country === 'USA';
+        const isMilitary = !!address.militaryPostOfficeTypeCode;
+        return isUSA || isMilitary;
+      },
+      'ui:options': {
+        labels: pciuStateLabels,
+        updateSchema: (formData, stateSchema) => {
+          /* eslint-disable no-param-reassign */
+          const address = formData.veteran[addressName];
+          const viewState = address['view:state'];
+          const state = address.state;
+          const militaryState = address.militaryState;
+          if (militaryStates.includes(viewState)) {
+            formData.veteran[addressName].militaryState = viewState;
+            delete formData.veteran[addressName].state;
+            delete formData.veteran[addressName].city;
+          } else if (viewState) {
+            formData.veteran[addressName].state = viewState;
+            delete formData.veteran[addressName].militaryState;
+            delete formData.veteran[addressName].militaryPostOfficeTypeCode;
+          } else {
+            formData.veteran[addressName]['view:state'] = (state || militaryState);
+          }
+          return stateSchema;
+          /* eslint-enable no-param-reassign */
+        }
+      }
+    },
+    state: {
+      'ui:options': {
+        labels: pciuStateLabels,
+        hideIf: () => true
+      }
+    },
+    addressLine1: {
+      'ui:title': 'Street',
+      'ui:validations': [validateSpecialCharacter],
```

Comment:
Can we stick with using `'ui:errorMessages': { pattern: 'validation message here' }`? I'm all for not running through another validation function if we don't have to.